### PR TITLE
Added GetNmStateList func to nmstate pkg

### PR DESCRIPTION
**Pull request by shaior**

### CODE REVIEW

File: pkg/nmstate/nmstate.go [hunk 1]
```diff
@@ -219,3 +219,11 @@ func (builder *Builder) validate() (bool, error) {
 
 	return true, nil
 }
+
+// GetNmStateList returns a NMState list.
+func GetNmStateList(apiClient *clients.Settings) (nmstateV1.NMStateList, error) {
```
[the remark this answers is not in the document]
Comment (shaior) [reply pkg/nmstate/nmstate.go]: implemented, thanks.

File: pkg/nmstate/list.go [hunk 2]
```diff
@@ -35,3 +37,30 @@ func ListPolicy(apiClient *clients.Settings) ([]*PolicyBuilder, error) {
 
 	return networkConfigurationPolicyObjects, nil
 }
+
+// ListNmStateConfig returns a NMStateConfig list.
+func ListNmStateConfig(apiClient *clients.Settings) ([]*NmStateConfigBuilder, error) {
+	nmStateConfigList := &assistedv1beta1.NMStateConfigList{}
+
+	err := apiClient.List(context.Background(), nmStateConfigList, &client.ListOptions{})
+	if err != nil {
```
Comment (kononovn) [inline pkg/nmstate/list.go:46]: Minor: Please add space.

File: pkg/nmstate/nmstateconfig.go [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+	// Created NMStateConfig object on the cluster.
+	Object *assistedv1beta1.NMStateConfig
+	// API client to interact with the cluster.
+	apiClient *clients.Settings
+	// errorMsg is processed before NMStateConfig object is created.
+	errorMsg string
+}
+
+// NewBuilder creates a new instance of NMStateConfig Builder.
+func NewNmStateConfigBuilder(apiClient *clients.Settings, name, namespace string) *NmStateConfigBuilder {
```
Comment (kononovn) [inline pkg/nmstate/nmstateconfig.go:31]: Please re-name to  NewConfigBuilder. You have nmstate in the name of the package. 
So if you need to trigger builder you'll use following call:
nmstate.NewConfigBuilder.

Comment (kononovn) [inline pkg/nmstate/nmstateconfig.go:31]: Have you tested this fucntion? Does it work?

File: pkg/nmstate/nmstateconfig.go [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+		apiClient: apiClient,
+		Definition: &assistedv1beta1.NMStateConfig{
+			ObjectMeta: metaV1.ObjectMeta{
+				Name:      name,
+				Namespace: namespace,
+			},
+		},
+	}
+
+	if name == "" {
```
Comment (kononovn) [inline pkg/nmstate/nmstateconfig.go:44]: Please add the same check for nmstate

**Review by kononovn — CHANGES_REQUESTED** [review]
Hi @shaior . 
Overall looks good. Just few more steps required.

According to import NMStateConfig is part of assisted-service:
github.com/openshift/assisted-service/api/v1beta1
Please move these files under **./pkg/asissted**

File: pkg/nmstate/nmstateconfig.go [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+	Definition *assistedv1beta1.NMStateConfig
+	// Created NMStateConfig object on the cluster.
+	Object *assistedv1beta1.NMStateConfig
+	// API client to interact with the cluster.
+	apiClient *clients.Settings
+	// errorMsg is processed before NMStateConfig object is created.
+	errorMsg string
+}
+
+// NewBuilder creates a new instance of NMStateConfig Builder.
```
Comment (trewest) [inline pkg/nmstate/nmstateconfig.go:30]:
```suggestion
// NewNmStateConfigBuilder creates a new instance of NMStateConfig Builder.
```

File: pkg/nmstate/nmstateconfig.go [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+func NewNmStateConfigBuilder(apiClient *clients.Settings, name, namespace string) *NmStateConfigBuilder {
+	glog.V(100).Infof("Initializing new NMStateConfig structure with the name: %s", name)
```
Comment (trewest) [inline pkg/nmstate/nmstateconfig.go:32]: Please include namespace in the glog

File: pkg/nmstate/nmstateconfig.go [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+	if valid, _ := builder.validate(); !valid {
+		return false
+	}
+
+	glog.V(100).Infof("Checking if NMStateConfig %s exists", builder.Definition.Name)
+
+	var err error
+	builder.Object, err = builder.Get()
+
+	if err != nil {
```
Comment (trewest) [inline pkg/nmstate/nmstateconfig.go:64]: I don't think we need a separate conditional for a glog. Error checking the return value of this method should provide why the error occurred

File: pkg/nmstate/nmstateconfig.go [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+	return nmStateConfig, err
+}
+
+// Create makes a NMStateConfig in the cluster and stores the created object in struct.
+func (builder *NmStateConfigBuilder) Create() (*NmStateConfigBuilder, error) {
+	if valid, err := builder.validate(); !valid {
+		return builder, err
+	}
+
+	glog.V(100).Infof("Creating the NMStateConfig %s", builder.Definition.Name)
```
Comment (trewest) [inline pkg/nmstate/nmstateconfig.go:100]: Please include the namespace

File: pkg/nmstate/nmstateconfig.go [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+	return builder, err
+}
+
+// Delete removes NMStateConfig object from a cluster.
+func (builder *NmStateConfigBuilder) Delete() (*NmStateConfigBuilder, error) {
+	if valid, err := builder.validate(); !valid {
+		return builder, err
+	}
+
+	glog.V(100).Infof("Deleting the NMStateConfig object %s", builder.Definition.Name)
```
Comment (trewest) [inline pkg/nmstate/nmstateconfig.go:119]: Please include namespace

File: pkg/nmstate/nmstateconfig.go [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+	return err == nil || !k8serrors.IsNotFound(err)
+}
+
+// Get returns NMStateConfig object if found.
+func (builder *NmStateConfigBuilder) Get() (*assistedv1beta1.NMStateConfig, error) {
+	if valid, err := builder.validate(); !valid {
+		return nil, err
+	}
+
+	glog.V(100).Infof("Collecting NMStateConfig object %s", builder.Definition.Name)
```
Comment (trewest) [inline pkg/nmstate/nmstateconfig.go:77]: Please include namespace

File: pkg/nmstate/nmstateconfig.go [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+	return &builder
+}
+
+// Exists checks whether the given NMStateConfig exists.
+func (builder *NmStateConfigBuilder) Exists() bool {
+	if valid, _ := builder.validate(); !valid {
+		return false
+	}
+
+	glog.V(100).Infof("Checking if NMStateConfig %s exists", builder.Definition.Name)
```
Comment (trewest) [inline pkg/nmstate/nmstateconfig.go:59]: Please include namespace

File: pkg/nmstate/list.go [hunk 12]
```diff
@@ -35,3 +37,30 @@ func ListPolicy(apiClient *clients.Settings) ([]*PolicyBuilder, error) {
 
 	return networkConfigurationPolicyObjects, nil
 }
+
+// ListNmStateConfig returns a NMStateConfig list.
+func ListNmStateConfig(apiClient *clients.Settings) ([]*NmStateConfigBuilder, error) {
```
Comment (trewest) [inline pkg/nmstate/list.go:42]:
```suggestion
func ListNmStateConfigs(apiClient *clients.Settings, namespace string) ([]*NmStateConfigBuilder, error) {
```
Please include a namespace parameter for this function. If you are trying to get all nmstateconfigs across all namespaces intentionally, please add a separate function called ListNmStateConfigsInAllNamespaces

File: pkg/nmstate/nmstateconfig.go [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+	"github.com/openshift-kni/eco-goinfra/pkg/clients"
+	"github.com/openshift-kni/eco-goinfra/pkg/msg"
+	assistedv1beta1 "github.com/openshift/assisted-service/api/v1beta1"
+	goclient "sigs.k8s.io/controller-runtime/pkg/client"
+
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metaV1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// Builder provides struct for the NMStateConfig object containing connection to
```
Comment (trewest) [inline pkg/nmstate/nmstateconfig.go:17]:
```suggestion
// NmStateConfigBuilder provides struct for the NMStateConfig object containing connection to
```

**Review by trewest — CHANGES_REQUESTED** [review]
Few comments

File: pkg/nmstate/nmstateconfig.go (hunk 3 above)
Comment (shaior) [reply pkg/nmstate/nmstateconfig.go:31]: yes.

File: pkg/nmstate/nmstateconfig.go (hunk 4 above)
Comment (shaior) [reply pkg/nmstate/nmstateconfig.go]: I assume you meant the namespace, and fixed it accordingly, if you meant something else, please clarify.
thanks.

File: test.go [hunk 14]
```diff
@@ -0,0 +1,26 @@
+package main
```
Comment (kononovn) [inline test.go:1]: Please remove this file from your PR

File: pkg/assisted/list.go [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+)
+
+// ListNmStateConfigsInAllNamespaces returns a a cluster-wide NMStateConfig list.
+func ListNmStateConfigsInAllNamespaces(apiClient *clients.Settings) ([]*NmStateConfigBuilder, error) {
+	nmStateConfigList := &assistedv1beta1.NMStateConfigList{}
+
+	err := apiClient.List(context.Background(), nmStateConfigList, &client.ListOptions{})
+
+	if err != nil {
+		glog.V(100).Infof("Failed to list nmStateConfigs due to %s", err.Error())
```
Comment (kononovn) [inline pkg/assisted/list.go:19]: Please update to something like:
glog.V(100).Infof("Failed to list nmStateConfigs across all namespaces due to %s", err.Error())

File: pkg/assisted/list.go [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+}
+
+// ListNmStateConfigs returns a NMStateConfig list in a given namespace.
+func ListNmStateConfigs(apiClient *clients.Settings, namespace string) ([]*NmStateConfigBuilder, error) {
+	nmStateConfigList := &assistedv1beta1.NMStateConfigList{}
+
+	err := apiClient.List(context.Background(), nmStateConfigList, &client.ListOptions{Namespace: namespace})
+
+	if err != nil {
+		glog.V(100).Infof("Failed to list nmStateConfig due to %s in namespace: %s",
```
Comment (kononovn) [inline pkg/assisted/list.go:47]: Please update to something like 
glog.V(100).Infof("Failed to list nmStateConfig in namespace: %s due to %s", err.Error(), namespace)

**Review by kononovn — CHANGES_REQUESTED** [review]
Please fix few comments above.

File: test.go (hunk 14 above)
Comment (shaior) [reply test.go:1]: Sure, it was added by mistake, thanks.

File: pkg/assisted/list.go [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+		nmstateConfigObjects = append(nmstateConfigObjects, nmStateConfBuilder)
+	}
+
+	return nmstateConfigObjects, err
+}
+
+// ListNmStateConfigs returns a NMStateConfig list in a given namespace.
+func ListNmStateConfigs(apiClient *clients.Settings, namespace string) ([]*NmStateConfigBuilder, error) {
+	nmStateConfigList := &assistedv1beta1.NMStateConfigList{}
+
```
Comment (trewest) [inline pkg/assisted/list.go:43]: Please validate that namespace is not an empty string

File: pkg/assisted/nmstateconfig.go [hunk 18]
```diff
@@ -0,0 +1,168 @@
...
+				Name:      name,
+				Namespace: namespace,
+			},
+		},
+	}
+
+	if name == "" {
+		glog.V(100).Infof("The name of the NMStateConfig is empty")
+
+		builder.errorMsg = "NMStateConfig 'name' cannot be empty"
```
Comment (trewest) [inline pkg/assisted/nmstateconfig.go:47]: Please use all lowercase characters

File: pkg/assisted/nmstateconfig.go [hunk 19]
```diff
@@ -0,0 +1,168 @@
...
+	}
+
+	if namespace == "" {
+		glog.V(100).Infof("The namespace of the NMStateConfig is empty")
+
+		builder.errorMsg = "NMStateConfig namespace's name is empty"
```
Comment (trewest) [inline pkg/assisted/nmstateconfig.go:53]: Please use all lowercase characters

File: pkg/assisted/list.go [hunk 20]
```diff
@@ -0,0 +1,67 @@
+package assisted
```
Comment (trewest) [inline pkg/assisted/list.go:1]: Can you move the functions in this file to nmstateconfig.go?

**Review by trewest — CHANGES_REQUESTED** [review]
Few comments

**Review by kononovn — DISMISSED** [review]
lgtm

**Comment by kononovn** [comment]
@shaior lgtm. Please get 1 more from @trewest

File: pkg/assisted/nmstateconfig.go [hunk 21]
```diff
@@ -0,0 +1,229 @@
...
+	if builder.errorMsg != "" {
+		glog.V(100).Infof("The %s builder has error message: %s", resourceCRD, builder.errorMsg)
+
+		return false, fmt.Errorf(builder.errorMsg)
+	}
+
+	return true, nil
+}
+
+// ListNmStateConfigsInAllNamespaces returns a a cluster-wide NMStateConfig list.
```
Comment (trewest) [inline pkg/assisted/nmstateconfig.go:170]:
```suggestion
// ListNmStateConfigsInAllNamespaces returns a cluster-wide NMStateConfig list.
```

File: pkg/assisted/nmstateconfig.go [hunk 22]
```diff
@@ -0,0 +1,229 @@
...
+	}
+
+	builder.Object = nil
+
+	return builder, nil
+}
+
+// validate will check that the builder and builder definition are properly initialized before
+// accessing any member fields.
+func (builder *NmStateConfigBuilder) validate() (bool, error) {
```
Comment (trewest) [inline pkg/assisted/nmstateconfig.go:140]: Please ensure that private functions and methods are listed at the bottom of the file

**Review by trewest — CHANGES_REQUESTED** [review]
Looks good, just two minor fixes and we can merge

**Review by trewest — APPROVED** [review]
/lgtm

**Pull request merged**